File: py/behavioral/memento.py
"""

Memento Pattern

Majorly used to build record:undo/roll-back cases.

1. Originator itself handles the creation of a snapshot and how
   to restore it.
2. Consumer/History-tracker/Caretaker of Memento/Snapshot consumes through an
   simplified interface without much access to internal state of the Memento.
3. Both Originator or Memento can define the restore func. If the later one, it
   simply passes to the Originator to execute restore behavior.


NOTE(shengzhc):
To make Memento encapsulated, the class can be privately defined in Originator,
and Originator class/module only exposes interface out.

"""


class IMemento:
	def __init__(self, originator):
		self.originator = originator

	def restore(self):
		if self.originator is not None:
			self.originator.restore(self)
		else:
			raise Exception("Originator is None unexpectedly")


class ConcreteMemento(IMemento):
	def __init__(self, originator, states):
		super().__init__(originator)
		self.states = states


class Originator:
	def __init__(self):
		self.__counter = 3

	def save(self):
		print("Saving state, before: ", self.__counter, " and after: ", self.__counter + 1)
		states = range(self.__counter)
		self.__counter += 1
		return ConcreteMemento(self, states)

	def restore(self, memento):
		prev = self.__counter
		self.__counter = len(memento.states)
		print("Restoring state, before: ", prev, " and after: ", self.__counter)


def client():
	o = Originator()
	history = []
	history.append(o.save())
	history.append(o.save())
	history.append(o.save())
	for m in history[::-1]:
		m.restore()


if __name__ == "__main__":
	client()

File: c++/src/behavior/memento.cpp
#include <iostream>

#include "memento.h"

namespace memento {
	using namespace std;
	Memento *Originator::save() {
		return new Memento(value + 5);
	}

	void Originator::restore(Memento *m) {
		this->value = m->value - 5;
	}

	void Originator::onDesc() {
		cout << "Originator's current value: " << value << endl;
	}

	void Originator::doSomeWork() {
	 	value ++;
	}

	void CareTaker::save() {
		Memento *const m = this->originator->save();
		this->list.push_back(m);
	}

	void CareTaker::undo() {
		Memento *const m = this->list.back();
		this->list.pop_back();
		this->originator->restore(m);
	}

	void MementoDesignPattern::example() {
		Originator *const o = new Originator();
		CareTaker *const ct = new CareTaker(o);
		o->doSomeWork();
		ct->save();
		cout << "After 1st save:" << endl;
		o->onDesc();
		o->doSomeWork();
		ct->save();
		cout << "After 2nd save:" << endl;
		o->onDesc();
		ct->undo();
		cout << "After 1st undo:" << endl;
		o->onDesc();
		ct->undo();
		cout << "Aftere 2nd undo:" << endl;
		o->onDesc();
	}

	string MementoDesignPattern::getName() {
		return "Memento Design Pattern";
	}

}